CONTRIBUTING.md

```diff
@@ -0,0 +1,48 @@
+# How to contribute
+
+## Submitting PRs
+
+Please follow the provided PR template when writing a description for your changes.
+
+Do not take criticism personally. When you get feedback, it is about your work,
+not your character, personality, etc. Keep in mind we all want to make the project better.
+
+When something is not clear, please ask questions to clear things up.
+
+If you would like to introduce a big architectural changes or do a refactoring
+either in the codebase or the development tools, please open an issue with a proposal
+first. This way we can think together about the problem and probably come up
+with a better solution.
+
+## Coding conventions and guidelines
+
+### Commit messages
+
+* Always write descriptive commit messages ("fix bug" is not acceptable).
+* Use the present tense ("Add feature" not "Added feature").
+* Use the imperative mood ("Move cursor to..." not "Moves cursor to...").
+* Limit the first line to 72 characters or less.
+* Include the number of the issue you are fixing.
+
+### Coding guidelines
+
+As a Python project, we must follow PEP8 guidelines.
+
+Furthermore, follow the Clean code conventions. The most important
+in this project are the following rules:
+
+* Simpler is always better.
```

Comment:
> Simpler is always better.

I would remove the word "always" and add a link to the KISS principle.

Reply:
In the world of _python zen_ (see below) there is the rule `Explicit is better than implicit.`

WEB development (and search) is implemented much more implicit. Can we give some words at hand, where it is a good choice to follow python zen and where we better be implicit .. 

This is often my problem, I can write very small patches but I often asking myself when it is better to implement a abstraction layer and design the code more modular.  My experience is, that it is most often better if I implement layers and make the code modular.   This can not be done in small patches --> In practice I really have a problem with the word "Simple" in "Simple is better" .. 

@asciimoo do you have some thoughts about?

----

````
 $ python3 -c "import this"
The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
````

Reply:
We can refer to this list. But I would not copy the whole thing over here as users can read it in other web pages.

Reply:
Sorry if I was misleading:  please do not copy the list .. copy nothing from the list, its just a reference I gave at hand.

My subject was on "Simple is always better" and what does it mean.  Does "to be simple" mean "to be implicit", does it mean "not to implement infrastructure code" ..  does it mean "do not modularize your code" and  "do not have indirectives " .. where does "Simple" becomes "bad" (does "Simple" violate other rules like single-point of definition)?

Reply:
>  My experience is, that it is most often better if I implement layers and make the code modular.

There  is no good way to handle this topic in general. It's very case specific.
First of all, modularity and implementing abstraction layers are two different things. You can achieve modularity without abstractions. Especially in a high level programming language like python. Unnecessary abstraction layers can be as bad as a highly coupled code and it can really slow down the development and make reviews/refactors painful and error prone.

Reply:
> can be as bad as a highly coupled code and it can really slow down the development and make reviews/refactors painful and error prone.

I think I'm starting to understand what you mean: what you described in this sentence is what I know from paid jobs. Academical abstraction layers conceived by big teams of consultant. Often seen in Java, SOAP, XML projects and other "enterprise software". / No I want this either.

At the moment, however, I am still a little unsettled, because I also see that my PRs do not quite meet expectations. / I hope I find the right balance.